Typechecker fails to resolve methods for impl blocks as statements

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
fn main() {
    struct Foo;

    trait Bar {
        fn foo(&self);
    }

    impl Bar for Foo {
        fn foo(&self) {}
    }

    let s = Foo;
    s.foo();
}
```

I expected to see this happen: no error

Instead, this happened:

```rust
test.rs:13:7: error: failed to resolve method for ‘foo’
   13 |     s.foo();
      |       ^~~
test.rs:13:5: error: failed to type resolve expression
   13 |     s.foo();
      |     ^
```

### Meta

- What version of Rust GCC were you using, git sha if possible.
058682f736b

--- a/gcc/rust/backend/rust-compile-expr.cc
+++ b/gcc/rust/backend/rust-compile-expr.cc
@@ -1698,23 +1698,6 @@ CompileExpr::visit (HIR::MethodCallExpr &expr)
   // method receiver
   tree self = CompileExpr::Compile (expr.get_receiver ().get (), ctx);
 
-  // lookup the resolved name
-  NodeId resolved_node_id = UNKNOWN_NODEID;
-  if (!ctx->get_resolver ()->lookup_resolved_name (
-	expr.get_mappings ().get_nodeid (), &resolved_node_id))
-    {
-      rust_error_at (expr.get_locus (), "failed to lookup resolved MethodCall");
-      return;
-    }
-
-  // reverse lookup
-  HirId ref;
-  if (!ctx->get_mappings ()->lookup_node_to_hir (resolved_node_id, &ref))
-    {
-      rust_fatal_error (expr.get_locus (), "reverse lookup failure");
-      return;
-    }
-
   // lookup the expected function type
   TyTy::BaseType *lookup_fntype = nullptr;
   bool ok = ctx->get_tyctx ()->lookup_type (
@@ -1871,20 +1854,10 @@ CompileExpr::resolve_operator_overload (
     expr.get_mappings ().get_hirid (), &fntype);
   rust_assert (is_op_overload);
 
-  // lookup the resolved name
-  NodeId resolved_node_id = UNKNOWN_NODEID;
-  bool ok = ctx->get_resolver ()->lookup_resolved_name (
-    expr.get_mappings ().get_nodeid (), &resolved_node_id);
-  rust_assert (ok);
-
-  // reverse lookup
-  HirId ref;
-  ok = ctx->get_mappings ()->lookup_node_to_hir (resolved_node_id, &ref);
-  rust_assert (ok);
-
   TyTy::BaseType *receiver = nullptr;
-  ok = ctx->get_tyctx ()->lookup_receiver (expr.get_mappings ().get_hirid (),
-					   &receiver);
+  bool ok
+    = ctx->get_tyctx ()->lookup_receiver (expr.get_mappings ().get_hirid (),
+					  &receiver);
   rust_assert (ok);
 
   bool is_generic_receiver = receiver->get_kind () == TyTy::TypeKind::PARAM;

--- a/gcc/rust/hir/rust-ast-lower-stmt.cc
+++ b/gcc/rust/hir/rust-ast-lower-stmt.cc
@@ -16,10 +16,9 @@
 // along with GCC; see the file COPYING3.  If not see
 // <http://www.gnu.org/licenses/>.
 
+#include "rust-ast-lower-item.h"
 #include "rust-ast-lower-stmt.h"
-#include "rust-ast-lower-enumitem.h"
 #include "rust-ast-lower-type.h"
-#include "rust-ast-lower-block.h"
 #include "rust-ast-lower-expr.h"
 #include "rust-ast-lower-pattern.h"
 
@@ -40,16 +39,6 @@ ASTLoweringStmt::translate (AST::Stmt *stmt, bool *terminated)
     resolver.translated->get_mappings ().get_hirid (),
     resolver.translated->get_locus ());
   resolver.mappings->insert_hir_stmt (resolver.translated);
-  if (resolver.translated->is_item ())
-    {
-      HIR::Item *i = static_cast<HIR::Item *> (resolver.translated);
-
-      auto defid = resolver.translated->get_mappings ().get_defid ();
-
-      resolver.handle_outer_attributes (*i);
-      resolver.mappings->insert_hir_item (i);
-      resolver.mappings->insert_defid_mapping (defid, i);
-    }
 
   return resolver.translated;
 }
@@ -72,21 +61,7 @@ ASTLoweringStmt::visit (AST::ExprStmt &stmt)
 void
 ASTLoweringStmt::visit (AST::ConstantItem &constant)
 {
-  HIR::Visibility vis = translate_visibility (constant.get_visibility ());
-
-  HIR::Type *type = ASTLoweringType::translate (constant.get_type ().get ());
-  HIR::Expr *expr = ASTLoweringExpr::translate (constant.get_expr ().get ());
-
-  auto crate_num = mappings->get_current_crate ();
-  Analysis::NodeMapping mapping (crate_num, constant.get_node_id (),
-				 mappings->get_next_hir_id (crate_num),
-				 mappings->get_next_localdef_id (crate_num));
-
-  translated = new HIR::ConstantItem (mapping, constant.get_identifier (), vis,
-				      std::unique_ptr<HIR::Type> (type),
-				      std::unique_ptr<HIR::Expr> (expr),
-				      constant.get_outer_attrs (),
-				      constant.get_locus ());
+  translated = ASTLoweringItem::translate (&constant);
 }
 
 void
@@ -116,185 +91,25 @@ ASTLoweringStmt::visit (AST::LetStmt &stmt)
 void
 ASTLoweringStmt::visit (AST::TupleStruct &struct_decl)
 {
-  std::vector<std::unique_ptr<HIR::GenericParam>> generic_params;
-  if (struct_decl.has_generics ())
-    {
-      generic_params = lower_generic_params (struct_decl.get_generic_params ());
-    }
-
-  std::vector<std::unique_ptr<HIR::WhereClauseItem>> where_clause_items;
-  HIR::WhereClause where_clause (std::move (where_clause_items));
-  HIR::Visibility vis = translate_visibility (struct_decl.get_visibility ());
-
-  std::vector<HIR::TupleField> fields;
-  for (AST::TupleField &field : struct_decl.get_fields ())
-    {
-      HIR::Visibility vis = translate_visibility (field.get_visibility ());
-      HIR::Type *type
-	= ASTLoweringType::translate (field.get_field_type ().get ());
-
-      auto crate_num = mappings->get_current_crate ();
-      Analysis::NodeMapping mapping (crate_num, field.get_node_id (),
-				     mappings->get_next_hir_id (crate_num),
-				     mappings->get_next_localdef_id (
-				       crate_num));
-
-      HIR::TupleField translated_field (mapping,
-					std::unique_ptr<HIR::Type> (type), vis,
-					field.get_locus (),
-					field.get_outer_attrs ());
-      fields.push_back (std::move (translated_field));
-    }
-
-  auto crate_num = mappings->get_current_crate ();
-  Analysis::NodeMapping mapping (crate_num, struct_decl.get_node_id (),
-				 mappings->get_next_hir_id (crate_num),
-				 mappings->get_next_localdef_id (crate_num));
-
-  translated = new HIR::TupleStruct (mapping, std::move (fields),
-				     struct_decl.get_identifier (),
-				     std::move (generic_params),
-				     std::move (where_clause), vis,
-				     struct_decl.get_outer_attrs (),
-				     struct_decl.get_locus ());
+  translated = ASTLoweringItem::translate (&struct_decl);
 }
 
 void
 ASTLoweringStmt::visit (AST::StructStruct &struct_decl)
 {
-  std::vector<std::unique_ptr<HIR::GenericParam>> generic_params;
-  if (struct_decl.has_generics ())
-    {
-      generic_params = lower_generic_params (struct_decl.get_generic_params ());
-    }
-
-  std::vector<std::unique_ptr<HIR::WhereClauseItem>> where_clause_items;
-  HIR::WhereClause where_clause (std::move (where_clause_items));
-  HIR::Visibility vis = translate_visibility (struct_decl.get_visibility ());
-
-  bool is_unit = struct_decl.is_unit_struct ();
-  std::vector<HIR::StructField> fields;
-  for (AST::StructField &field : struct_decl.get_fields ())
-    {
-      HIR::Visibility vis = translate_visibility (field.get_visibility ());
-      HIR::Type *type
-	= ASTLoweringType::translate (field.get_field_type ().get ());
-
-      auto crate_num = mappings->get_current_crate ();
-      Analysis::NodeMapping mapping (crate_num, field.get_node_id (),
-				     mappings->get_next_hir_id (crate_num),
-				     mappings->get_next_localdef_id (
-				       crate_num));
-
-      HIR::StructField translated_field (mapping, field.get_field_name (),
-					 std::unique_ptr<HIR::Type> (type), vis,
-					 field.get_locus (),
-					 field.get_outer_attrs ());
-
-      if (struct_field_name_exists (fields, translated_field))
-	break;
-
-      fields.push_back (std::move (translated_field));
-    }
-
-  auto crate_num = mappings->get_current_crate ();
-  Analysis::NodeMapping mapping (crate_num, struct_decl.get_node_id (),
-				 mappings->get_next_hir_id (crate_num),
-				 mappings->get_next_localdef_id (crate_num));
-
-  translated = new HIR::StructStruct (mapping, std::move (fields),
-				      struct_decl.get_identifier (),
-				      std::move (generic_params),
-				      std::move (where_clause), is_unit, vis,
-				      struct_decl.get_outer_attrs (),
-				      struct_decl.get_locus ());
+  translated = ASTLoweringItem::translate (&struct_decl);
 }
 
 void
 ASTLoweringStmt::visit (AST::Union &union_decl)
 {
-  std::vector<std::unique_ptr<HIR::GenericParam>> generic_params;
-  if (union_decl.has_generics ())
-    {
-      generic_params = lower_generic_params (union_decl.get_generic_params ());
-    }
-
-  std::vector<std::unique_ptr<HIR::WhereClauseItem>> where_clause_items;
-  HIR::WhereClause where_clause (std::move (where_clause_items));
-  HIR::Visibility vis = translate_visibility (union_decl.get_visibility ());
-
-  std::vector<HIR::StructField> variants;
-  for (AST::StructField &variant : union_decl.get_variants ())
-    {
-      HIR::Visibility vis = translate_visibility (variant.get_visibility ());
-      HIR::Type *type
-	= ASTLoweringType::translate (variant.get_field_type ().get ());
-
-      auto crate_num = mappings->get_current_crate ();
-      Analysis::NodeMapping mapping (crate_num, variant.get_node_id (),
-				     mappings->get_next_hir_id (crate_num),
-				     mappings->get_next_localdef_id (
-				       crate_num));
-
-      HIR::StructField translated_variant (mapping, variant.get_field_name (),
-					   std::unique_ptr<HIR::Type> (type),
-					   vis, variant.get_locus (),
-					   variant.get_outer_attrs ());
-
-      if (struct_field_name_exists (variants, translated_variant))
-	break;
-
-      variants.push_back (std::move (translated_variant));
-    }
-
-  auto crate_num = mappings->get_current_crate ();
-  Analysis::NodeMapping mapping (crate_num, union_decl.get_node_id (),
-				 mappings->get_next_hir_id (crate_num),
-				 mappings->get_next_localdef_id (crate_num));
-
-  translated
-    = new HIR::Union (mapping, union_decl.get_identifier (), vis,
-		      std::move (generic_params), std::move (where_clause),
-		      std::move (variants), union_decl.get_outer_attrs (),
-		      union_decl.get_locus ());
+  translated = ASTLoweringItem::translate (&union_decl);
 }
 
 void
 ASTLoweringStmt::visit (AST::Enum &enum_decl)
 {
-  std::vector<std::unique_ptr<HIR::GenericParam>> generic_params;
-  if (enum_decl.has_generics ())
-    {
-      generic_params = lower_generic_params (enum_decl.get_generic_params ());
-    }
-
-  std::vector<std::unique_ptr<HIR::WhereClauseItem>> where_clause_items;
-  HIR::WhereClause where_clause (std::move (where_clause_items));
-  HIR::Visibility vis = translate_visibility (enum_decl.get_visibility ());
-
-  // bool is_unit = enum_decl.is_zero_variant ();
-  std::vector<std::unique_ptr<HIR::EnumItem>> items;
-  for (auto &variant : enum_decl.get_variants ())
-    {
-      HIR::EnumItem *hir_item = ASTLoweringEnumItem::translate (variant.get ());
-      items.push_back (std::unique_ptr<HIR::EnumItem> (hir_item));
-    }
-
-  auto crate_num = mappings->get_current_crate ();
-  Analysis::NodeMapping mapping (crate_num, enum_decl.get_node_id (),
-				 mappings->get_next_hir_id (crate_num),
-				 mappings->get_next_localdef_id (crate_num));
-
-  HIR::Enum *hir_enum
-    = new HIR::Enum (mapping, enum_decl.get_identifier (), vis,
-		     std::move (generic_params), std::move (where_clause),
-		     std::move (items), enum_decl.get_outer_attrs (),
-		     enum_decl.get_locus ());
-  translated = hir_enum;
-  for (auto &variant : hir_enum->get_variants ())
-    {
-      mappings->insert_hir_enumitem (hir_enum, variant.get ());
-    }
+  translated = ASTLoweringItem::translate (&enum_decl);
 }
 
 void
@@ -311,77 +126,7 @@ ASTLoweringStmt::visit (AST::EmptyStmt &empty)
 void
 ASTLoweringStmt::visit (AST::Function &function)
 {
-  // ignore for now and leave empty
-  std::vector<std::unique_ptr<HIR::WhereClauseItem>> where_clause_items;
-  HIR::WhereClause where_clause (std::move (where_clause_items));
-  HIR::FunctionQualifiers qualifiers
-    = lower_qualifiers (function.get_qualifiers ());
-  HIR::Visibility vis = translate_visibility (function.get_visibility ());
-
-  // need
-  std::vector<std::unique_ptr<HIR::GenericParam>> generic_params;
-  if (function.has_generics ())
-    {
-      generic_params = lower_generic_params (function.get_generic_params ());
-    }
-
-  Identifier function_name = function.get_function_name ();
-  Location locus = function.get_locus ();
-
-  std::unique_ptr<HIR::Type> return_type
-    = function.has_return_type () ? std::unique_ptr<HIR::Type> (
-	ASTLoweringType::translate (function.get_return_type ().get ()))
-				  : nullptr;
-
-  std::vector<HIR::FunctionParam> function_params;
-  for (auto &param : function.get_function_params ())
-    {
-      auto translated_pattern = std::unique_ptr<HIR::Pattern> (
-	ASTLoweringPattern::translate (param.get_pattern ().get ()));
-      auto translated_type = std::unique_ptr<HIR::Type> (
-	ASTLoweringType::translate (param.get_type ().get ()));
-
-      auto crate_num = mappings->get_current_crate ();
-      Analysis::NodeMapping mapping (crate_num, param.get_node_id (),
-				     mappings->get_next_hir_id (crate_num),
-				     UNKNOWN_LOCAL_DEFID);
-
-      auto hir_param
-	= HIR::FunctionParam (mapping, std::move (translated_pattern),
-			      std::move (translated_type), param.get_locus ());
-      function_params.push_back (hir_param);
-    }
-
-  bool terminated = false;
-  std::unique_ptr<HIR::BlockExpr> function_body
-    = std::unique_ptr<HIR::BlockExpr> (
-      ASTLoweringBlock::translate (function.get_definition ().get (),
-				   &terminated));
-
-  auto crate_num = mappings->get_current_crate ();
-  Analysis::NodeMapping mapping (crate_num, function.get_node_id (),
-				 mappings->get_next_hir_id (crate_num),
-				 mappings->get_next_localdef_id (crate_num));
-
-  mappings->insert_location (function_body->get_mappings ().get_hirid (),
-			     function.get_locus ());
-
-  auto fn
-    = new HIR::Function (mapping, std::move (function_name),
-			 std::move (qualifiers), std::move (generic_params),
-			 std::move (function_params), std::move (return_type),
-			 std::move (where_clause), std::move (function_body),
-			 std::move (vis), function.get_outer_attrs (),
-			 HIR::SelfParam::error (), locus);
-
-  // add the mappings for the function params at the end
-  for (auto &param : fn->get_function_params ())
-    {
-      mappings->insert_hir_param (&param);
-      mappings->insert_location (mapping.get_hirid (), param.get_locus ());
-    }
-
-  translated = fn;
+  translated = ASTLoweringItem::translate (&function);
 }
 
 void
@@ -394,6 +139,24 @@ void
 ASTLoweringStmt::visit (AST::MacroRulesDefinition &def)
 {
   lower_macro_definition (def);
+}
+
+void
+ASTLoweringStmt::visit (AST::Trait &trait)
+{
+  translated = ASTLoweringItem::translate (&trait);
+}
+
+void
+ASTLoweringStmt::visit (AST::InherentImpl &impl_block)
+{
+  translated = ASTLoweringItem::translate (&impl_block);
+}
+
+void
+ASTLoweringStmt::visit (AST::TraitImpl &impl_block)
+{
+  translated = ASTLoweringItem::translate (&impl_block);
 }
 
 } // namespace HIR

--- a/gcc/rust/hir/rust-ast-lower-stmt.h
+++ b/gcc/rust/hir/rust-ast-lower-stmt.h
@@ -42,6 +42,9 @@ public:
   void visit (AST::Function &function) override;
   void visit (AST::ExternBlock &extern_block) override;
   void visit (AST::MacroRulesDefinition &extern_block) override;
+  void visit (AST::Trait &trait) override;
+  void visit (AST::InherentImpl &impl_block) override;
+  void visit (AST::TraitImpl &impl_block) override;
 
 private:
   ASTLoweringStmt () : translated (nullptr), terminated (false) {}

--- a/gcc/rust/resolve/rust-ast-resolve-item.cc
+++ b/gcc/rust/resolve/rust-ast-resolve-item.cc
@@ -699,8 +699,10 @@ ResolveItem::visit (AST::TraitImpl &impl_block)
 
   resolver->get_name_scope ().push (scope_node_id);
   resolver->get_type_scope ().push (scope_node_id);
+  resolver->get_label_scope ().push (scope_node_id);
   resolver->push_new_name_rib (resolver->get_name_scope ().peek ());
   resolver->push_new_type_rib (resolver->get_type_scope ().peek ());
+  resolver->push_new_label_rib (resolver->get_type_scope ().peek ());
 
   if (impl_block.has_generics ())
     for (auto &generic : impl_block.get_generic_params ())
@@ -714,8 +716,9 @@ ResolveItem::visit (AST::TraitImpl &impl_block)
   NodeId trait_resolved_node = ResolveType::go (&impl_block.get_trait_path ());
   if (trait_resolved_node == UNKNOWN_NODEID)
     {
-      resolver->get_type_scope ().pop ();
       resolver->get_name_scope ().pop ();
+      resolver->get_type_scope ().pop ();
+      resolver->get_label_scope ().pop ();
       return;
     }
 
@@ -723,8 +726,9 @@ ResolveItem::visit (AST::TraitImpl &impl_block)
   NodeId type_resolved_node = ResolveType::go (impl_block.get_type ().get ());
   if (type_resolved_node == UNKNOWN_NODEID)
     {
-      resolver->get_type_scope ().pop ();
       resolver->get_name_scope ().pop ();
+      resolver->get_type_scope ().pop ();
+      resolver->get_label_scope ().pop ();
       return;
     }
 
@@ -792,9 +796,12 @@ ResolveItem::visit (AST::TraitImpl &impl_block)
       resolve_impl_item (impl_item.get (), impl_prefix, cpath);
     }
 
-  resolver->get_type_scope ().peek ()->clear_name (
-    Self, impl_block.get_type ()->get_node_id ());
+  Rib *r = resolver->get_type_scope ().peek ();
+  r->clear_name (Self, impl_block.get_type ()->get_node_id ());
+
+  resolver->get_name_scope ().pop ();
   resolver->get_type_scope ().pop ();
+  resolver->get_label_scope ().pop ();
 }
 
 void

--- a/gcc/rust/resolve/rust-ast-resolve-pattern.cc
+++ b/gcc/rust/resolve/rust-ast-resolve-pattern.cc
@@ -18,7 +18,6 @@
 
 #include "rust-ast-resolve-pattern.h"
 #include "rust-ast-resolve-path.h"
-#include "rust-ast-resolve-expr.h"
 
 namespace Rust {
 namespace Resolver {

--- a/gcc/rust/resolve/rust-ast-resolve-stmt.cc
+++ b/gcc/rust/resolve/rust-ast-resolve-stmt.cc
@@ -16,6 +16,7 @@
 // along with GCC; see the file COPYING3.  If not see
 // <http://www.gnu.org/licenses/>.
 
+#include "rust-ast-resolve-toplevel.h"
 #include "rust-ast-resolve-item.h"
 #include "rust-ast-resolve-stmt.h"
 #include "rust-ast-resolve-implitem.h"
@@ -33,6 +34,27 @@ ResolveStmt::visit (AST::ExternBlock &extern_block)
 				     CanonicalPath::create_empty ());
       ResolveExternItem::go (item.get (), prefix, canonical_prefix);
     }
+}
+
+void
+ResolveStmt::visit (AST::Trait &trait)
+{
+  ResolveTopLevel::go (&trait, prefix, canonical_prefix);
+  ResolveItem::go (&trait, prefix, canonical_prefix);
+}
+
+void
+ResolveStmt::visit (AST::InherentImpl &impl_block)
+{
+  ResolveTopLevel::go (&impl_block, prefix, canonical_prefix);
+  ResolveItem::go (&impl_block, prefix, canonical_prefix);
+}
+
+void
+ResolveStmt::visit (AST::TraitImpl &impl_block)
+{
+  ResolveTopLevel::go (&impl_block, prefix, canonical_prefix);
+  ResolveItem::go (&impl_block, prefix, canonical_prefix);
 }
 
 } // namespace Resolver

--- a/gcc/rust/resolve/rust-ast-resolve-stmt.h
+++ b/gcc/rust/resolve/rust-ast-resolve-stmt.h
@@ -359,6 +359,9 @@ public:
   }
 
   void visit (AST::ExternBlock &extern_block) override;
+  void visit (AST::Trait &trait) override;
+  void visit (AST::InherentImpl &impl_block) override;
+  void visit (AST::TraitImpl &impl_block) override;
 
 private:
   ResolveStmt (const CanonicalPath &prefix,

--- a/gcc/rust/typecheck/rust-hir-type-check-expr.cc
+++ b/gcc/rust/typecheck/rust-hir-type-check-expr.cc
@@ -1207,8 +1207,16 @@ TypeCheckExpr::visit (HIR::MethodCallExpr &expr)
   context->insert_type (expr.get_method_name ().get_mappings (), lookup);
 
   // set up the resolved name on the path
-  resolver->insert_resolved_name (expr.get_mappings ().get_nodeid (),
-				  resolved_node_id);
+  if (resolver->get_name_scope ().decl_was_declared_here (resolved_node_id))
+    {
+      resolver->insert_resolved_name (expr.get_mappings ().get_nodeid (),
+				      resolved_node_id);
+    }
+  else
+    {
+      resolver->insert_resolved_misc (expr.get_mappings ().get_nodeid (),
+				      resolved_node_id);
+    }
 
   // return the result of the function back
   infered = function_ret_tyty;

--- a/gcc/testsuite/rust/compile/issue-2238.rs
+++ b/gcc/testsuite/rust/compile/issue-2238.rs
@@ -1,0 +1,15 @@
+fn main() {
+    struct Foo;
+
+    trait Bar {
+        fn foo(&self);
+    }
+
+    impl Bar for Foo {
+        fn foo(&self) {}
+        // { dg-warning "unused name" "" { target *-*-* } .-1 }
+    }
+
+    let s = Foo;
+    s.foo();
+}

